rbac/api/cross_access/serializer.py

```diff
@@ -108,25 +109,21 @@ def get_roles(self, obj):
 
     def create(self, validated_data):
         """Override the create method to associate the roles to cross account request after it is created."""
-        role_data = validated_data.pop("roles")
-        display_names = [role["display_name"] for role in role_data]
+        validated_data.pop("roles")
         request = CrossAccountRequest.objects.create(**validated_data)
         cross_account_access_handler(request, self.context["user"])
-
-        roles = Role.objects.filter(display_name__in=display_names)
-        for role in roles:
-            request.roles.add(role)
+        role_uuids = [role["uuid"] for role in self.context["request"].data["roles"]]
+        request.roles.add(*role_uuids)
```

Comment:
I was talking with @astrozzc about this during collab. I wonder if we could move the validation and transformation to the serializer, if that would break things?

Something like this?

```
    def validate_roles(self, roles):
        public_tenant = Tenant.objects.get(tenant_name="public")
        for role_name in roles:
            try:
                role = Role.objects.get(tenant=public_tenant, display_name=role_name)
                if not role.system:
                    raise serializers.ValidationError(
                        "Only system roles may be assigned to a cross-account-request.", code="cross-account-request"
                    )
            except Role.DoesNotExist:
                raise serializers.ValidationError(f"Role '{role_name}' does not exist.", code="cross-account-request")
        return [{"uuid": role.uuid} for role in roles]

    def create(self, validated_data):
        """Override the create method to associate the roles to cross account request after it is created."""
        role_data = validated_data.pop("roles")
        role_uuids = [role["uuid"] for role in role_data]
        request = CrossAccountRequest.objects.create(**validated_data)
        cross_account_access_handler(request, self.context["user"])
        request.roles.add(*role_uuids)
```

And then you would take out the format / find role by display name stuff in the view and just leave it to the serializer.

This feels like what Django Rest Framework _wants_ you to do, but dunno if it would change the existing behavior. Not tested!!

Reply:
Reference: https://www.django-rest-framework.org/api-guide/serializers/#field-level-validation

Reply:
Tried this would work: the to_internal_value will call the role serializer validation, if we don't transform the display_name (strings) to list of {"uuid": "xxx"} before that, it would raise validation error.
```
    def to_internal_value(self, data):
        public_tenant = Tenant.objects.get(tenant_name="public")
        roles = data["roles"]
        role_objs = Role.objects.filter(tenant=public_tenant, display_name__in=roles)
        role_names = {role_obj.display_name for role_obj in role_objs}
        for role_name in roles:
            if role_name not in role_names:
                raise serializers.ValidationError(f"Role `'{role_name}'` does not exist.", code="cross-account-request")
        data["roles"] = [{"uuid": role.uuid} for role in role_objs]
        return super().to_internal_value(data)
```

Reply:
I wonder, why does returning the value from the validate role method not work? It looked like it used that value for the validated data from the docs & code

Reply:
Yea, I spent sometime tracking it down, because it runs the validation of the RoleSerializer first before running the validate_roles of the CrossAccountDetailSerializer. The RoleSerializer requires it to be a dictionary like {"display_names": "xxx"} instead of list of display_names ["xxx"]